api/src/main/java/org/geysermc/geyser/api/event/lifecycle/GeyserDefineCustomBlocksEvent.java

```diff
@@ -0,0 +1,66 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.api.event.lifecycle;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.geysermc.geyser.api.block.custom.CustomBlockData;
+import org.geysermc.geyser.api.block.custom.CustomBlockState;
+import org.geysermc.event.Event;
+
+/**
+ * Called on Geyser's startup when looking for custom blocks. Custom blocks must be registered through this event.
+ *
+ * This event will not be called if the "add-non-bedrock-items" setting is disabled in the Geyser config.
+ */
+public abstract class GeyserDefineCustomBlocksEvent implements Event {
+
+    /**
+     * Registers the given {@link CustomBlockData} as a custom block
+     *
+     * @param customBlockData the custom block to register
+     */
+    public abstract void registerCustomBlock(@NonNull CustomBlockData customBlockData);
+
+    /**
+     * Registers the given {@link CustomBlockState} as an override for the
+     * given java state identifier
+     * Java state identifiers are listed in
+     * https://raw.githubusercontent.com/GeyserMC/mappings/master/blocks.json
+     *
+     * @param javaIdentifier the java state identifier to override
+     * @param customBlockState the custom block state with which to override java state identifier
+     */
+    public abstract void registerBlockStateOverride(@NonNull String javaIdentifier, @NonNull CustomBlockState customBlockState);
+
+    /**
+     * Registers the given {@link CustomBlockData} as an override for the
+     * given java item identifier
+     *
+     * @param javaIdentifier the java item identifier to override
+     * @param customBlockData the custom block data with which to override java item identifier
+     */
+    public abstract void registerBlockItemOverride(@NonNull String javaIdentifier, @NonNull CustomBlockData customBlockData);
```

Comment:
```suggestion
    public abstract void registerItemOverride(@NonNull String javaIdentifier, @NonNull CustomBlockData customBlockData);
```

May also be worth adding a variant that takes in a `CustomItemData` instance.

Reply:
https://github.com/GeyserMC/Geyser/pull/3505/commits/410712cb78f1e7df00d9099b409664714fafec16#diff-5482ed5484e9ade47524ad19eeea5751d2de4d5940e1f8830447476ff6000be2L83-R83

In terms of allowing this to be used with custom items, this is meant to make the block show up for the item of a specific vanilla java identifier for inventory purposes (e.g. if you remodel the oak fence block you probably want the oak fence in your inventory to look like an oak fence, versus if you've remodeled a bunch of noteblock states you probably don't particularly care that the noteblock is still a noteblock). But for using custom blocks as a stand-in for custom items, I've come to the conclusion there are just too many adverse consequences that we could better address in other ways in the future (e.g. item frames as entities, custom geometry dropped item entities). But displaying a custom item as a block kinda just breaks a ton of things since blocks effectively have none of the components that items had, so they're good for little more than their utility as an inventory display (which they still effectively suck at due to https://bugs.mojang.com/browse/MCPE-123305)